### ISSUE TITLE
libappstream regression: Flathub generates invalid remote icon entry

### ISSUE THREAD

**Issue by sophie-h** [flathub/flathub#4222]
Bold claim in the issue title, but generating apps.gnome.org is broken since Jun 6, 2023 6:16am GMT+0200. The reason is an unexpected value in `type="remote"` for icons. 

> remote icons loaded from a remote URL. Currently, only HTTP/HTTPS urls are supported. This icon type should have width and height properties.

https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-icon

So [rust appstream](https://github.com/bilelmoussaoui/appstream) refuses the `app/drey/EarTag/...` value in the example below:

```xml
    <icon height="48" type="cached" width="48">app.drey.EarTag.png</icon>
    <icon height="64" type="cached" width="64">app.drey.EarTag.png</icon>
    <icon height="128" type="remote" width="128">app/drey/EarTag/48e5a929d7aad9d76df6deeb49cbf86f/icons/128x128/app.drey.EarTag.png</icon>
    <icon height="128" type="cached" width="128">app.drey.EarTag.png</icon>
    <icon type="stock">app.drey.EarTag</icon>
```

**Comment by razzeee** [comment]
They are relative now, there is a media base url (I'm actually not sure, it might get stripped right now) on one of the root elements

We handle this hardcoded on the website for now https://github.com/flathub/website/commit/f740030e8c890bd4316a7309fcc8dbd96046b617#diff-9985c8101c5c6fdd34d0bed35f279f01d429eee5f739521c86e114335c3e1fcbR64

**Comment by sophie-h** [comment]
I think it's missing in here. (The version number also seems sus.)

```xml
<components version="0.8" origin="flatpak">
```

**Comment by razzeee** [comment]
Bart is aware, but there's a state holiday and he has a vacation day tomorrow

**Comment by mcrha** [comment]
gnome-software bug https://gitlab.gnome.org/GNOME/gnome-software/-/issues/2196

The paths look odd too, and they care not valid with added `https://dl.flathub.org/` prefix to them (using `wget`).

The same applies also to `<screenshots>`.

**Comment by razzeee** [comment]
The base path should be `https://dl.flathub.org/media` and not just `https://dl.flathub.org`

**Comment by mcrha** [comment]
I see. That explains why it did not work for me. It was only a guess from my side.

**Comment by ximion** [comment]
If flatpak-builder or whatever runs `appstreamcli compose` passes the `--no-partial-urls` flag, all URLs will be complete and the media-baseurl will not be used, for increase backwards compatibility. That could help here too!
(this was a feature that was previously not exposed on the command line, but is available with AppStream 0.16.3).

**Comment by bbhtt** [comment]
This should be solved by https://github.com/flathub/org.flatpak.Builder/commit/3bacdbbddaecbc61fb3614be5f083d7ce603b603

@sophie-h can you retrigger a build, so that the catalogue data gets composed with libappstream now and check?

**Comment by sophie-h** [comment]
> Everything should be relative to the mirror URL `https://dl.flathub.org/media/`

Hm, why should they be relative if the linked patch adds `--no-partial-urls`?

**Comment by bbhtt** [comment]
I meant to say they will be complete URLs starting with https://dl.flathub.org/media/

**Comment by sophie-h** [comment]
Yes, seems to work. The only app with a relative URL is a leftover from the previous transition to libappstream

```
WARN apps_gnome_org::generate: URL parser error: relative URL without a base
 | <component type="desktop">
 |   <id>org.lyx.LyX</id>
 |   <translation type="gettext">LyX</translation>
 |   <name>LyX</name>
 |   <summary>The Document Processor</summary>
 |   <developer_name>LyX</developer_name>
 |  …
 ```

**Comment by bbhtt** [comment]
Where?

**Comment by bbhtt** [comment]
> In the "list" view, in Gnome Software (45.3), on apps like Kdenlive, Discord or Extension manager (on the detailed app page, the icon shows fine).

Can you open a separate issue for this? This is unrelated to the issue here.

It seems libappstream is giving only 128px now:

```
ls app-info/media/com/discordapp/Discord.desktop/2c3067076b8497e454766b79db8589e6/icons/        
 128x128
```

The list view depends on a 64px icon. cc @ximion

**Comment by ximion** [comment]
> The list view depends on a 64px icon. @ximion, any idea what to do here?

Do you apply any odd configuration to appstreamcli? What is its full command-line?

**Comment by bbhtt** [comment]
> > The list view depends on a 64px icon. @ximion, any idea what to do here?
> 
> Do you apply any odd configuration to appstreamcli? What is its full command-line?

https://github.com/flatpak/flatpak-builder/blob/8c036e00630e35423c03388aacc06cd00dda74ea/src/builder-manifest.c#L3052-L3086

```
appstreamcli compose --no-partial-urls --prefix=/ --origin=com.discordapp.Discord --media-baseurl=https://dl.flathub.org/media/ --media-dir=/home/bbhtt/Downloads/com.discordapp.Discord/.flatpak-builder/rofiles/rofiles-7yoATF/files/share/app-info/media --result-root=/home/bbhtt/Downloads/com.discordapp.Discord/.flatpak-builder/rofiles/rofiles-7yoATF/files --data-dir=/home/bbhtt/Downloads/com.discordapp.Discord/.flatpak-builder/rofiles/rofiles-7yoATF/files/share/app-info/xmls --icons-dir=/home/bbhtt/Downloads/com.discordapp.Discord/.flatpak-builder/rofiles/rofiles-7yoATF/files/share/app-info/icons/flatpak '--components=com.discordapp.Discord,com.discordapp.Discord.desktop' /home/bbhtt/Downloads/com.discordapp.Discord/.flatpak-builder/rofiles/rofiles-7yoATF/files
```

**Comment by ximion** [comment]
Okay, that command-line looks good. A 64x64px icon should *always* be generated unconditionally, if it isn't there, compose should fail. So I guess the next thing would be to examine the output of that command (not sure how I could test that, but maybe I could have a look later, maybe tomorrow).

**Comment by bbhtt** [comment]
> Okay, that command-line looks good. A 64x64px icon should _always_ be generated unconditionally, if it isn't there, compose should fail. So I guess the next thing would be to examine the output of that command (not sure how I could test that, but maybe I could have a look later, maybe tomorrow).

You can manually run the compose step. This should be roughly equivalent.

```patch
diff --git a/com.discordapp.Discord.json b/com.discordapp.Discord.json
index 6c9dd79..bbada2d 100644
--- a/com.discordapp.Discord.json
+++ b/com.discordapp.Discord.json
@@ -7,6 +7,7 @@
     "sdk": "org.freedesktop.Sdk",
     "command": "com.discordapp.Discord",
     "separate-locales": false,
+    "appstream-compose": false,
     "tags": [
         "proprietary"
     ],
@@ -62,9 +63,12 @@
                 "install -d /app/share/applications",
                 "sed -e 's/Icon=discord/Icon=com.discordapp.Discord/' -e 's|Exec=/usr/share/discord/Discord|Exec=com.discordapp.Discord|' /app/discord/discord.desktop > /app/share/applications/${FLATPAK_ID}.desktop",
                 "install -Dm644 /app/discord/discord.png /app/share/icons/hicolor/256x256/apps/${FLATPAK_ID}.png",
-                "install -Dm644 com.discordapp.Discord.appdata.xml /app/share/appdata/com.discordapp.Discord.appdata.xml",
+                "install -Dm644 com.discordapp.Discord.appdata.xml /app/share/metainfo/com.discordapp.Discord.appdata.xml",
                 "patch-desktop-filename ${FLATPAK_DEST}/discord/resources/app.asar"
             ],
+            "post-install": [
+                "appstreamcli compose --verbose --no-partial-urls --prefix=/ --origin=${FLATPAK_ID} --media-baseurl=https://dl.flathub.org/media/ --media-dir=${FLATPAK_DEST}/share/app-info/media --result-root=${FLATPAK_DEST} --data-dir=${FLATPAK_DEST}/share/app-info/xmls --icons-dir=${FLATPAK_DEST}/share/app-info/icons/flatpak --components=${FLATPAK_ID}.desktop ${FLATPAK_DEST}"
+            ],
             "sources": [
                 {
                     "type": "archive",
```

It seems to be generated but not put into the `media` directory.

**Comment by ximion** [comment]
> It seems to be generated but not put into `app-info/media/com/discordapp/Discord.desktop/<hash>/icons/` like 128px. Is 64px supposed to be cached only?

That is expected, I think - only stuff that is to be served from a remote location ends up in `media`.
The 64x64px icon is supposed to be cache-only, the AppStream spec mandates that this size must always be available, so serving it as remote icon makes no sense if we will always have it on disk.

**Comment by bbhtt** [comment]
So one issues is, appstream is generating the cached icon entry as <icon type="cached" width="64" height="64">com.discordapp.Discord.desktop.png</icon> with a `.desktop` suffix but flatpak is committing them with the `$FLATPAK_ID` instead `com.discordapp.Discord` in the appstream ref

https://dl.flathub.org/repo/appstream/x86_64/icons/64x64/com.discordapp.Discord.png is there but https://dl.flathub.org/repo/appstream/x86_64/icons/64x64/com.discordapp.Discord.desktop.png is not.

I'm not sure why flatpak's appstream cache is missing `com.discordapp.Discord.png` even though it is on the server.

**Comment by bbhtt** [comment]
So seems like it is hitting https://github.com/flatpak/flatpak/blob/f9cbfe1fd67b608b0df1f3f1c333e2b972fda225/common/flatpak-utils.c#L5311-L5312 then https://github.com/flatpak/flatpak/blob/f9cbfe1fd67b608b0df1f3f1c333e2b972fda225/common/flatpak-utils.c#L5209-L5210 because that's what I get when generating the appstream ref.

I guess Flathub has it in some kind of cache but `flatpak update` downloads the latest commit of the ref which doesn't have it.

I guess appstream-glib ignored the ID and generated icons with `$FLATPAK_ID` and everything used to work.

**Comment by bbhtt** [comment]
This seems hard to solve, maybe applications can change the `id` tag in metainfo file and add the replaces stuff and everything should be back to normal?

There is an apparent disconnection between `$FLATPAK_ID` and `<id>` which breaks things.

**Comment by ximion** [comment]
Yeah, compose will *always* use the current component-ID for the icon name, verbatim without changing it. We had some issues in the past when compose tried to be a bit too clever with it, so just using the name was the result :-)

Why doesn't the builder tool enforce that $FLATPAK_ID == appstream-component-ID?

**Comment by bbhtt** [comment]
> Why doesn't the builder tool enforce that $FLATPAK_ID == appstream-component-ID?

Seems for compat https://github.com/flatpak/flatpak-builder/commit/7292b2d63f7125727b0a7692c9d693dee24addbf

**Comment by bbhtt** [comment]
I made an attempt to fix discord in https://github.com/flathub/com.discordapp.Discord/pull/384, I don't know what else to do here, since both tools are correct in their own right.

Unfortunately some new applications got in with `.desktop` suffix in appstream `id` and we can't block it in the linter due to historical baggage.

**Comment by mcrha** [comment]
> maybe applications can change the `id` tag in metainfo file and add the replaces stuff and everything should be back to normal

That would be a very bad workaround, forcing everyone to touch their apps, not talking you cannot just change the ID, as is explained and discouraged in the lengthy **Note** of the ID tag definition https://www.freedesktop.org/software/appstream/docs/sect-Metadata-Application.html#tag-id-desktopapp

**Comment by bbhtt** [comment]
What do you suggest to do here? I know about the warning, but the replaces tag is also for cases like this https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-replaces

**Comment by mcrha** [comment]
For example gnome-software does not know about "replaces" tag. I've no idea what other appstream data consumers do.

Still, changing the app id is a bad idea, you break everything with it, from the connection between different repos (flatpak, PackageKit, rpm-ostree,... all should use the same app ID, to be considered as alternatives; gnome-software constantly fights between similar inconsistencies, where an app with `.desktop` suffix does not match the same app without the `.desktop` suffix in the id from another repository). There is also a question about app reviews, which are by ID - when you change the ID, the reviews are effectively gone.

You wrote above that there is `64x64/com.discordapp.Discord.png`, but not `64x64/com.discordapp.Discord.desktop.png`, and the appstream data references `com.discordapp.Discord.desktop.png`. A logic change would be to correct the generated appstream data to have`<icon .../>` without that `.desktop` in the name, when the files are generated that way these days.

**Comment by bbhtt** [comment]
Further complicating things is flatpak-builder is supplying both `$FLATPAK_ID` and `$FLATPAK_ID.desktop` to appstream-compose, appstream compose doesn't know which one is the actual `$FLATPAK_ID` defined in the manifest.

**Comment by ximion** [comment]
TBH, I think we should at least look into whether we could resolve this for good to make `$FLATPAK_ID == appstream-component-ID` a reality. Is there any way to check which applications have this mismatch and how many there are?
If it's just a few, we can probably fix them and remove the workaround, which would save us from so much pain in the long run (especially with apps that have names like [org.telegram.desktop](https://flathub.org/apps/org.telegram.desktop) where the `.desktop` is **not** an added suffix but an intentional part of the rDNS ID). If it is too many apps to fix or if it's too widespread, then we need to think of other solutions.

Renaming the component-ID and adding `replaces` is not pretty, but it could be a price worth paying in this case... Or renaming the Flatpak ID, but I guess that would break updates, which would be a lot worse.

**Comment by mcrha** [comment]
I'm sorry, I still do not understand why you expect apps to "fix" something caused by Flathub.org due to using different appstream tool to generate its own data.

If the `<icon>` entry in the appstream data provided by Flathub.org is generated by this appstream tool, or added by the Flathub.org or flatpak itself, then these tools are supposed to be fixed, not the apps.

A very lame approach to get apps with `.desktop` suffix ID:
```
cat /var/lib/flatpak/appstream/flathub/x86_64/active/appstream.xml | grep -C2 "<component" | grep "<id>" | grep ".desktop<" | wc -l
390
```

with `sort -u` it's 387, out of:

```
$ cat /var/lib/flatpak/appstream/flathub/x86_64/active/appstream.xml | grep "<component" | wc -l
3648
```

components.

**Comment by ximion** [comment]
Stripping the desktop-suffix or mangling the component-ID in any way is very wrong, and caused us a lot of pain in the past on the AppStream side.

> A very lame approach to get apps with .desktop suffix ID

Could we limit them somehow to apps that also have a different FLATPAK_ID? This for certain has many false-positives...

What would speak against adding a workaround to the builder to just commit the icon with `$FLATPAK_ID` and `$FLATPAK_ID.desktop`, and ensure at least one of them is present? That should resolve this, right?
(Adding a Flatpak-specific workaround to AppStream if `$FLATPAK_ID` is in the environment is also possible, but I would *really* like to avoid that... Data generation is already pretty complex, adding a varying ID would make it much worse).

**Comment by ramcq** [comment]
This seems like it's become overcomplex somehow. I like the idea of Flatpak just providing the flatpak ID to appstream - that seems like the right thing to do. Is it really problematic that within that context, appstreamcli could match either the (legacy) suffixed .desktop ID as well as the verbatim one?

Flatpak had that problem for a while that with im.telegram.desktop or whatever it would assume the ID was im.telegram, but this doesn't actually cause a real-world confusion (there is no im.telegram which is aliased) so we were able to fix it - as long as you're careful not to chop .desktop off mindlessly, could we make appstream just be a little more relaxed and pick up the files/icons/whatever named either id or id.desktop (provided id doesn't end in .desktop already) to account for the legacy behaviour in appstream? (might be able to limit such behaviour based on the declared version in the metainfo file?)

**Comment by ximion** [comment]
> This seems like it's become overcomplex somehow. I like the idea of Flatpak just providing the flatpak ID to appstream - that seems like the right thing to do.

I don't think so - why should AppStream act differently compared to anywhere else based on whether Flatpak has invoked it?

> Is it really problematic that within that context, appstreamcli could match either the (legacy) suffixed .desktop ID as well as the verbatim one?

It actually is - AppStream internally relies on the component-ID being consistent and unchanging (during the compose process). It is used in various places to uniquely identify auxiliary data, mangling it would cause a lot of potential breakage (I can't tell what exactly will break, but since we would violate an assumption, potentially a lot).
There is one place where we do use heuristics, and that is for `launchable`, which is less critical. And even those heuristics have caused some unforeseen issues in the past.

I've been burned by this a few times now, so if at all possible, I *really* want to avoid repeating that mistake again and adding too much magic to this process, especially conditional magic for Flatpak-only.

> Flatpak had that problem for a while that with im.telegram.desktop or whatever it would assume the ID was im.telegram, but this doesn't actually cause a real-world confusion (there is no im.telegram which is aliased) so we were able to fix it - as long as you're careful not to chop .desktop off mindlessly, could we make appstream just be a little more relaxed and pick up the files/icons/whatever named either id or id.desktop (provided id doesn't end in .desktop already) to account for the legacy behaviour in appstream? (might be able to limit such behaviour based on the declared version in the metainfo file?)

As far as I understand it, AppStream does behave correctly here and generates the expected data, but it is flatpak-builder which assumes some mangling has happened and does not commit an existing icon.
If AppStream was creating invalid data or wouldn't find an icon that was found before, then it would clearly be an AppStream bug, but that - as far as I understand the issue - is not the case here.
(To find icons, AppStream already has an incredible amount of heuristics to account for a ton of cases)

As far as I understand the problem, making flatpak-builder commit the icon with `$FLATPAK_ID` and `$FLATPAK_ID.desktop` instead of just assuming that the component-ID was mangled would resolve this issue. The builder could even be smart about it and check on its own if there is an ID mismatch and then adjust accordingly. E.g.:
```
if (FLATPAK_ID == COMPONENT_ID)
   commit_icon(FLATPAK_ID + ".png")
else if (FLATPAK_ID + ".desktop" == COMPONENT_ID)
  commit_icon(COMPONENT_ID + ".png")
else
  raise_error("Icon not found!")
``` 
(if I understand the cause of this issue correctly)

**Comment by bbhtt** [comment]
This is the list of affected apps:

<details>

<summary> Apps </summary>

```
app.lith.Lith
ar.com.softwareperonista.Rockarrolla
art.taunoerik.tauno-monitor
br.app.pw3270.terminal
ca._0ldsk00l.Nestopia
ca.desrt.dconf-editor
cc.arduino.arduinoide
com.albiononline.AlbionOnline
com.anydesk.Anydesk
com.bixense.PasswordCalculator
com.bladecoder.adventure-editor
com.chatterino.chatterino
com.diffingo.fwbackups
com.discordapp.Discord
com.dosbox.DOSBox
com.doycho.euterpe.gtk
com.dropbox.Client
com.ekonomikas.merkato
com.elsevier.MendeleyDesktop
com.endlessm.photos
com.frac_tion.teleport
com.github.amikha1lov.RecApp
com.github.appadeia.Taigo
com.github.bajoja.indicator-kdeconnect
com.github.birros.WebArchives
com.github.bjaraujo.Bombermaaan
com.github.dariasteam.cowsrevenge
com.github.donadigo.appeditor
com.github.elth0r0.iqpuzzle
com.github.fabiocolacio.marker
com.github.hluk.copyq
com.github.huluti.Coulr
com.github.inercia.k3x
com.github.JannikHv.Gydl
com.github.kmwallio.thiefmd
com.github.liferooter.textpieces
com.github.maoschanz.DynamicWallpaperEditor
com.github.miguelmota.Cointop
com.github.mikacousin.olc
com.github.paolostivanin.OTPClient
com.github.quaternion
com.github.theironrobin.siglo
com.github.wwmm.easyeffects
com.github.wwmm.pulseeffects
com.github.Xenoveritas.abuse
com.gitlab.bitseater.meteo
com.gitlab.coringao.JAG
com.google.AndroidStudio
com.googleplaymusicdesktopplayer.GPMDP
com.hack_computer.Clubhouse
com.hack_computer.OperatingSystemApp
com.hack_computer.Sidetrack
com.jagex.RuneScape
com.kavilgroup.gelectrical
com.kavilgroup.gestimator
com.leinardi.gkraken
com.leinardi.gst
com.leinardi.gwe
com.leinardi.gx52
com.lettier.movie-monad
com.mattjakeman.ExtensionManager
com.ozmartians.VidCutter
com.realm667.Wolfenstein_Blade_of_Agony
com.rosegardenmusic.rosegarden
com.skype.Client
com.sourcepole.kadas
com.teeworlds.Teeworlds
com.toggl.TogglDesktop
com.ultimaker.cura
com.uploadedlobster.peek
com.viber.Viber
com.vixalien.decibels
com.wings3d.WINGS
com.zquestclassic.ZQuest
de.gunibert.Hackgregator
de.philippun1.Snoop
dev.jamiethalacker.window_painter
de.wwwtech.ColorMate
de.zwarf.picplanner
edu.stanford.Almond
eu.ad5001.LogarithmPlotter
id.sideka.App
im.gitter.Gitter
info.bibletime.BibleTime
info.olasagasti.revelation
io.atom.Atom
io.botfather.Botfather
io.exodus.Exodus
io.github.aerocyber.sitemarker
io.github.amit9838.mousam
io.github.betaflight.BetaflightConfigurator
io.github.dida_code.OpstaKultura
io.github.diegoivanme.flowtime
io.github.diegoivan.pdf_metadata_editor
io.github.diegopvlk.Dosage
io.github.dyegoaurelio.simple-wireplumber-gui
io.github.FailurePoint.RandomNumberFive
io.github.fizzyizzy05.binary
io.github.flattool.Warehouse
io.github.hakandundar34coding.mini-system-monitor
io.github.hakandundar34coding.system-monitoring-center
io.github.ImEditor
io.github.leonardschardijn.Chirurgien
io.github.limads.Queries
io.github.lxndr.gswatcher
io.github.martinrotter.textosaurus
io.github.mki1967.mki3dgame
io.github.mmstick.FontFinder
io.github.nate_xyz.Chromatic
io.github.nate_xyz.Conjure
io.github.nate_xyz.Paleta
io.github.nate_xyz.Resonance
io.github.NhekoReborn.Nheko
io.github.nokse22.asciidraw
io.github.nokse22.inspector
io.github.nokse22.minitext
io.github.nokse22.teleprompter
io.github.nokse22.trivia-quiz
io.github.nokse22.ultimate-tic-tac-toe
io.github.orontee.Argos
io.github.phastmike.tags
io.github.qwersyk.Newelle
io.github.Rirusha.Cassette
io.github.RodZill4.Material-Maker
io.github.thiefmd.themegenerator
io.github.unrud.RecentFilter
io.gitlab.celleron56.libretile
io.gitlab.gregorni.Calligraphy
io.gitlab.gregorni.Letterpress
io.gitlab.gwendalj.package-transporter
io.gitlab.Turtlico
io.liri.Calculator
io.mgba.mGBA
io.otsaloma.gaupol
io.otsaloma.nfoview
it.mijorus.collector
it.mijorus.smile
it.mijorus.whisper
me.hergert.Schemes
me.iepure.devtoolbox
me.sanchezrodriguez.passes
net.bartkessels.getit
net.danigm.loop
net.lugsole.bible_gui
net.mediaarea.AVIMetaEdit
net.mediaarea.DVAnalyzer
net.mediaarea.MediaConch
net.mediaarea.MOVMetaEdit
net.mediaarea.QCTools
net.meijn.onvifviewer
net.oz9aec.Gpredict
net.redeclipse.RedEclipse
net.runelite.RuneLite
net.sf.fuse_emulator
net.sourceforge.Chessx
net.sourceforge.DuneLegacy
net.sourceforge.electrip.Electrip
net.sourceforge.Fillets
net.sourceforge.projectM
net.sourceforge.Teo
net.veloren.veloren
org.adishatz.syncpasswd
org.apache.netbeans
org.chrisoft.qmidiplayer
org.davmail.DavMail
org.develz.Crawl
org.eclipse.iot.fourdiac.Ide
org.emilien.Password
org.emilien.SpaceLaunch
org.fedoraproject.MediaWriter
org.freefilesync.FreeFileSync
org.fritzing.Fritzing
org.gabmus.unifydmin
org.gahshomar.Gahshomar
org.geogebra.GeoGebra
org.gnode.NixView
org.gnome.Boxes
org.gnome.Builder
org.gnome.Calculator
org.gnome.Calendar
org.gnome.Characters
org.gnome.Cheese
org.gnome.ColorViewer
org.gnome.Connections
org.gnome.Contacts
org.gnome.Devhelp
org.gnome.dspy
org.gnome.eog
org.gnome.Epiphany
org.gnome.Evolution
org.gnome.frogr
org.gnome.gbrainy
org.gnome.Geary
org.gnome.gedit
org.gnome.gitlab.bazylevnik0.Convolution
org.gnome.gitlab.Cowsay
org.gnome.gitlab.johannesjh.favagtk
org.gnome.gThumb
org.gnome.HexGL
org.gnome.Loupe
org.gnome.meld
org.gnome.moserial
org.gnome.Music
org.gnome.NetworkDisplays
org.gnome.Nibbles
org.gnome.Notes
org.gnome.OfficeRunner
org.gnome.Polari
org.gnome.PowerStats
org.gnome.Recipes
org.gnome.Rhythmbox3
org.gnome.Shotwell
org.gnome.SoundJuicer
org.gnome.SoundRecorder
org.gnome.SwellFoop
org.gnome.TextEditor
org.gnome.Totem
org.godotengine.Godot
org.kde.akregator
org.kde.ark
org.kde.artikulate
org.kde.blinken
org.kde.bomber
org.kde.bovo
org.kde.cantor
org.kde.choqok
org.kde.digikam
org.kde.dolphin
org.kde.elisa
org.kde.falkon
org.kde.filelight
org.kde.gcompris
org.kde.granatier
org.kde.gwenview
org.kde.index
org.kde.isoimagewriter
org.kde.juk
org.kde.kaffeine
org.kde.kalzium
org.kde.kamoso
org.kde.kanagram
org.kde.kapman
org.kde.katomic
org.kde.kbibtex
org.kde.kblackbox
org.kde.kblocks
org.kde.kbounce
org.kde.kbreakout
org.kde.kbruch
org.kde.kcachegrind
org.kde.kcalc
org.kde.kcolorchooser
org.kde.kdenlive
org.kde.kdevelop
org.kde.kdiamond
org.kde.kdiff3
org.kde.kfind
org.kde.kfourinline
org.kde.kgeography
org.kde.kgoldrunner
org.kde.kgraphviewer
org.kde.khangman
org.kde.kid3
org.kde.kig
org.kde.kigo
org.kde.kimagemapeditor
org.kde.kiriki
org.kde.kiten
org.kde.kjumpingcube
org.kde.kleopatra
org.kde.klettres
org.kde.klickety
org.kde.kmahjongg
org.kde.kmines
org.kde.kmplot
org.kde.kmymoney
org.kde.knavalbattle
org.kde.knetwalk
org.kde.knights
org.kde.koko
org.kde.kolf
org.kde.kollision
org.kde.kolourpaint
org.kde.kommit
org.kde.kompare
org.kde.kongress
org.kde.konquest
org.kde.konsole
org.kde.kontact
org.kde.konversation
org.kde.krdc
org.kde.krename
org.kde.kreversi
org.kde.kronometer
org.kde.kruler
org.kde.ksirk
org.kde.ksnakeduel
org.kde.ksquares
org.kde.kstars
org.kde.ksudoku
org.kde.kteatime
org.kde.ktimetracker
org.kde.ktorrent
org.kde.ktouch
org.kde.ktuberling
org.kde.kturtle
org.kde.kubrick
org.kde.kwordquiz
org.kde.kwrite
org.kde.kxstitch
org.kde.labplot2
org.kde.lokalize
org.kde.lskat
org.kde.marble
org.kde.massif-visualizer
org.kde.minuet
org.kde.okteta
org.kde.okular
org.kde.palapeli
org.kde.parley
org.kde.peruse
org.kde.picmi
org.kde.pix
org.kde.rocs
org.kde.ruqola
org.kde.skanpage
org.kde.SymbolEditor
org.kde.tellico
org.kde.umbrello
org.kde.vvave
org.kde.yakuake
org.keepassxc.KeePassXC
org.KekikAkademi.BTKSorgu
org.kekikakademi.eFatura
org.kekikakademi.ntvHaber
org.libreoffice.LibreOffice
org.libsdl.Maelstrom
org.mapeditor.Tiled
org.neverball.Neverball
org.nickvision.cavalier
org.octave.Octave
org.openclonk.OpenClonk
org.openfodder.OpenFodder
org.openmw.OpenMW
org.openttd.OpenTTD
org.parlatype.Parlatype
org.processing.processingide
org.pyzo.pyzo
org.qgis.qgis
org.quassel_irc.QuasselClient
org.quetoo.Quetoo
org.raceintospace.Raceintospace
org.remmina.Remmina
org.sauerbraten.Sauerbraten
org.shadered.SHADERed
org.solarus_games.solarus.Launcher
org.sparkleshare.SparkleShare
org.tabos.maxcontrol
org.tabos.saldo
org.thonny.Thonny
org.vim.Vim
org.worldpossible.ScriptLauncher
org.xonotic.Xonotic
page.codeberg.Imaginer.Imaginer
pm.mirko.Atoms
re.sonny.Retro
se.manyver.Manyverse
us.zoom.Zoom
work.openpaper.Paperwork
ws.openarena.OpenArena
xyz.slothlife.Jogger
```

</details>

**Comment by mcrha** [comment]
> This is the list of affected apps:

which is 370, +/- 1.

**Comment by ximion** [comment]
> which is 370, +/- 1.

Brr... That's an amount that would definitely warrant some kind of workaround, fixing that for every app would be a significant effort :-/

All of this really should be a different bug though, the issue here is actually something completely different that has been fixed already.

**Comment by hfiguiere** [comment]
you should know that NOW flatpak-builder does fix the app-id on `rename-appdata-file` or `rename-desktop-file`. It only did it for the latter if it didn't match the desktop-file.

**Comment by ximion** [comment]
Renaming the component-ID is pretty evil - but if we could do it only once to get rid of this discrepancy, I'd be all for it!
It just feels like that with this many affected apps, unless there is a way to automate it and ensure nothing breaks, we can't really do this and need some automated workaround...

**Comment by hfiguiere** [comment]
As I mentionned in a different channel, there were 2022 blog post reviewed by people in the know that still used the `.desktop` in app-id. And the fact that the templates used that, and that flathub explicitely allowed it didn't help.

The reason I implemented the rename in flatpak-builder is because otherwise it's very painfull to fix it as there is not tooling to "edit" the appstream file. (appstream-glib has a "modify" that would work for this, but it's being phased out).

**Comment by ximion** [comment]
I'd be fine with adding a feature to edit MetaInfo files to `appstreamcli`, but if it can be avoided not for component-IDs... Changing them is an antifeature, as so much stuff depends on them. I could see a feature to change the ID and place the previous ID in a `<replaces/>` block though, that's a supported pattern. And I'd rather do that than injecting component-ID heuristics into the compose process, if it helps.

But is there anything that speaks against the solution outlined at the very bottom of https://github.com/flathub/flathub/issues/4222#issuecomment-2020038035? Does Flatpak maybe require the icon name to match the FLATPAK_ID? (that would require us to search for a new solution - I do have some more ideas then though ^^).

**Comment by TingPing** [comment]
> Does Flatpak maybe require the icon name to match the FLATPAK_ID? 

Icons can match `$FLATPAK_ID` or `$FLATPAK_ID.*`. So `$FLATPAK_ID.desktop.png` is totally valid.

**Comment by ximion** [comment]
> Icons can match `$FLATPAK_ID` or `$FLATPAK_ID.*`. So `$FLATPAK_ID.desktop.png` is totally valid.

In that case, and if I understand the problem correctly, this may actually be pretty trivial to resolve :-)

**Comment by bbhtt** [comment]
What is stopping from dropping this line https://github.com/flatpak/flatpak/blob/f1088e3013adc5010012e16d060178b0f1d48226/common/flatpak-utils.c#L5311-L5312 to fix the issue? I don't think it regresses anything and will fix the issue here.

This was introduced for an even rarer appstream ID ie. two `.desktop`-s `com.telegram.desktop.desktop` https://github.com/flatpak/flatpak/commit/7dd92d8a9be2e14313e2cda7dea44298fbb005a4 Telegram has since then changed the appstream ID at some point and now it is only `com.telegram.desktop`.

In case of two `.desktops`-s appstreamcli produces the icons as `com.telegram.desktop.desktop.png` which would again hit the same issue here, but it is already so very rare and I think we can add a check in linter to make sure no such ID goes through on Flathub.

I have yet to see a two `.desktop` component ID in the appstream data.

**Comment by bbhtt** [comment]
PRs opened, the flatpak PR has the repro steps. I'll wait for others to chime in.

**Comment by ximion** [comment]
Thanks a lot! :-) Having just one ID for an app will help us greatly in future, I believe.

**Comment by bbhtt** [comment]
Flathub should no longer accept `id` tags with mismatches, changes have been deployed with exceptions for all existing apps and apps in the submission queue.

What remains is getting one of the Flatpak patches merged, getting it backported to Flatpak 1.14.x, getting a new release and getting a PPA build - then it can be deployed to the Flathub servers running Ubuntu.

**Comment by bbhtt** [comment]
Closing this. I didn't find any maintainer to get the Flatpak patch reviewed in an year and I lost interest, meanwhile in that time, app maintainers or app's upstreams started dropping the old `.desktop` suffix for whatever reason. So the count of affected apps have dropped from 370 to 256 right now.

If that's the way this gets solved then so be it. The linter check exists so new ones with mismatches won't get in. We can downstream patch it in Flathub repo server, if it continues to be a headache but so far no one has complained outside this issue.
